### PR TITLE
[ic-data-tables]: Truncation and text wrap functionality

### PR DESCRIPTION
**Pull request by jd239**
## Summary of the changes
Truncation and text wrap functionality added to data tables. The new functionality includes:
- Updates to `ic-data-table`
- New Cypress tests
- New VR tests
- Updates to typography and tooltip
- CSS data table updates
- Updated storybook
- Updated types

## Related issue
#1003 

### General 

- [x] Changes to docs package checked and committed.
- [x] All acceptance criteria reviewed and met. 

### Testing

- [x] Relevant unit tests and visual regression tests added. 
- [x] Visual testing against Figma component specification completed. 
- [x] Playground stories in React Storybook up to date, with any prop changes and additions addressed. 

### Accessibility 

- [x] Accessibility Insights FastPass performed.
- [x] A11y unit test added and yields no issues.
- [x] A11y plug-in on Storybook yields no issues. 
- [ ] Manual screen reader testing performed using NVDA and VoiceOver. 
- [x] Manual keyboard testing for keyboard controls and logical focus order. 
- [x] Correct roles used and ARIA attributes used correctly where required. 
- [x] Logical heading structure is maintained, and the HTML elements used for headings can be changed to fit within the wider page structure. 

### Resize/zoom behaviour 

- [x] Page can be zoomed to 400% with no loss of content. 
- [x] Screen magnifier used with no issues. 
- [x] Text resized to 200% with no loss of content.
- [x] Text spacing increased as per the [WCAG 1.4.12 success criterion](https://www.w3.org/TR/WCAG21/#text-spacing) with no loss of content.

### System modes

- [x] Browser setting 'prefers reduced motion' tested. No animations or motion visible whilst this setting is on. 
- [x] Windows High Contrast mode tested with no loss of content.
- [x] System light and dark mode tested with no loss of content.
- [x] Browser support tested (Chrome, Safari, Firefox and Edge). 

### Testing content extremes

- [x] Min/max content examples tested with no loss of content or overflow. 
- [x] All prop combinations work without issue. 
- [x] Tested for FOUC (Flash of Unstyled Content) in both SSR (Server-Side Rendering) and SSG (Static Site Generation) settings.
- [x] Controlled and uncontrolled input components tested.

### CODE REVIEW

**Review by GCHQ-Developer-112 — CHANGES_REQUESTED** [review]
Hefty piece of work - well done! This was just the first round of reviewing. There's still more I need to check over but have a left a few comments in the meantime

**Comment by GCHQ-Developer-112** [comment]
Spotted some issues in Storybook

1. **Related to data table with pagination:** The ellipsis on truncated text is no longer present if the number of items is more than 5 e.g. 10 or 15 items and you go forward to the next page and then backwards to the previous page

https://github.com/user-attachments/assets/d32cf273-0f2e-4dbe-a577-394fdf29b61f

2. **Related to showHide truncation pattern:** When the text has enough space to be shown in full, it is still truncated. There is also a delay between changing the row height and the "see more" button appearing. This is a bit of a worry for data-dense data tables.

https://github.com/user-attachments/assets/4f04466c-78ee-45e8-abac-73991e3d8537

File: packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx [hunk 1]
```diff
@@ -313,12 +679,70 @@ export class DataTable {
...
   }
 
   @Watch("globalRowHeight")
   @Watch("variableRowHeight")
   rowHeightChangeHandler(): void {
+    /**
+     * Is it better to regenerate the cells when the data object has been changed internally?
+     * This will make the data object the single source of truth however, regenerating cells
+     * when a change has been made could be memory intensive.
+     */
```
[the remark this answers is not in the document]
Comment (jd239) [reply packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx:738]: Removed

File: packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx [hunk 2]
```diff
@@ -375,6 +799,13 @@ export class DataTable {
     }
   };
 
+  private getCellOptions = (cell: any, key: string) => {
+    //TODO: Update function to accept value to replace getCellAlignment function
```
[the remark this answers is not in the document]
Comment (jd239) [reply packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx:803]: Nope. Removed.

**Comment by jd239** [comment]
>Related to showHide truncation pattern: When the text has enough space to be shown in full, it is still truncated. There is also a delay between changing the row height and the "see more" button appearing. This is a bit of a worry for data-dense data tables.

The issue with see more/see less is that I need the data to render before calculating whether it needs the show/hide button. Might be worth a discussion to figure out what is possible to make it a smoother experience.

**Review by ad9242 — CHANGES_REQUESTED** [review]

File: packages/canary-react/src/stories/ic-data-table.stories.mdx [hunk 3]
```diff
@@ -1284,6 +1286,180 @@ const DataTable = () => (
...
+### Truncating Cell data
+
+When a row is given a set height, content that no longer fits into the cell area, the content becomes truncated. There are two types of truncation:
+
+- `tooltip` [Default]: Adds a line-clamp to the cell and displays the message in a tooltip.
+- `showHide`: Adds a `See More/See Less` button underneath the cell text, allowing the user to display the full content should they wish.
+
+These values are set using the `truncationPattern` property.
+
+If there is a certain row or column that should be exempt from truncation, then a property can either be added to the row or the column object, named `textWrap`. Anything this is applied to will ignore a custom height and show the full content.
```
[the remark this answers is not in the document]
Comment (jd239) [reply packages/canary-react/src/stories/ic-data-table.stories.mdx:1298]: This has been amended and moved below text wrap

**Review by GCHQ-Developer-112 — APPROVED** [review]
As discussed

**Review by ad9242 — CHANGES_REQUESTED** [review]
There are some very high thresholds for some of the tests - some 30%+ & one is 49%. Could you just check over them, please?

File: packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx [hunk 4]
```diff
@@ -274,6 +633,27 @@ export class DataTable {
...
+  }: CustomEvent<{ value: number }>): void {
+    if ((target as HTMLIcPaginationBarElement).parentElement !== this.el) {
+      this.previousRowsPerPage = this.rowsPerPage;
+      this.rowsPerPage = detail.value;
+    }
+
+    this.itemsPerPageChange = true;
   }
 
   @Listen("icTableDensityUpdate")
```
[the remark this answers is not in the document]
Comment (jd239) [reply packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx:659]: This comes from Data table title bar: https://github.com/mi6/ic-ui-kit/blob/canary/data-tables-truncation-new/packages/canary-web-components/src/components/ic-data-table-title-bar/ic-data-table-title-bar.tsx#L58

File: packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx [hunk 5]
```diff
@@ -274,6 +633,27 @@ export class DataTable {
...
+    if (!this.initialLoad && this.previousPaginationPage !== detail.value) {
+      // This is to prevent icPageChange from triggering truncation on first load
+      this.icPageChangeEvent = true;
+    }
+
+    this.previousPaginationPage = detail.value;
+    this.initialLoad = false;
+  }
+
+  @Listen("icItemsPerPageChange")
```
[the remark this answers is not in the document]
Comment (jd239) [reply packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx:646]: This comes from the pagination-bar: https://github.com/mi6/ic-ui-kit/blob/canary/data-tables-truncation-new/packages/canary-web-components/src/components/ic-pagination-bar/ic-pagination-bar.tsx#L168

**Review by GCHQ-Developer-530 — CHANGES_REQUESTED** [review]
This looks really good! Just had a look through and the only thing I've noticed is that the see more/see less pattern can be a bit temperamental on smaller screens. The department column seems to work perfectly adding the ellipsis and the see more button stays inside the column width. But name and job title sometimes get cut off. 
Screenshot 2024-08-13 at 07 56 06
Screenshot 2024-08-13 at 07 55 49

File: packages/canary-react/cypress-image-diff-screenshots/baseline/IcDataTable.cy.tsx-density-dense.png
Comment (GCHQ-Developer-299) [inline packages/canary-react/cypress-image-diff-screenshots/baseline/IcDataTable.cy.tsx-density-dense.png:1]: The text in Last name overflows and is cut off - compared to the [Dense Storybook story](https://mi6.github.io/ic-ui-kit/branches/canary/data-tables-truncation-new/canary-react/?path=/story/react-components-data-table--dense) where the text wraps happily. Is the behaviour in this screenshot deliberate?

**Review by GCHQ-Developer-299 — DISMISSED** [review]

File: packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx [hunk 6]
```diff
@@ -187,6 +220,12 @@ export class DataTable {
    */
   @Prop() stickyRowHeaders?: boolean = false;
 
+  /**
+   * Sets the method used to truncate longer text in cells where textWrap is `false`.
+   * The `tooltip` truncation pattern allows the overflowing text to be seen in a tooltip. The `showHide` truncation pattern allows the overflowing text to be shown and hidden using the ic-typography "See more"/"See less" buttons.
```
Comment (ad9242) [inline packages/canary-web-components/src/components/ic-data-table/ic-data-table.tsx:225]:
```suggestion
   * The `tooltip` truncation pattern allows the overflowing text to be seen in a tooltip. The `show-hide` truncation pattern allows the overflowing text to be shown and hidden using the ic-typography "See more"/"See less" buttons.
```

string should be kebab-case

**Review by ad9242 — CHANGES_REQUESTED** [review]

**Review by GCHQ-Developer-530 — APPROVED** [review]

**Review by GCHQ-Developer-299 — APPROVED** [review]

**Review by ad9242 — APPROVED** [review]

**Pull request merged**